src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp

```diff
@@ -40,64 +40,41 @@
 // - Have fast comparisons
 // - Have constant time access
 // We achieve this by using a closed hashtable for finding previously existing NCS:s and referring to them by an index that's smaller than a pointer.
-class NativeCallStackStorage : public CHeapObj<mtNMT> {
+class NativeCallStackStorage : public CHeapObjBase {
 public:
   struct StackIndex {
     friend NativeCallStackStorage;
-
-  private:
-    static constexpr const int32_t _invalid = -1;
-
     int32_t _stack_index;
-    StackIndex(int32_t stack_index)
-      : _stack_index(stack_index) {
-    }
-
   public:
+    static constexpr const int32_t invalid = -1;
     static bool equals(const StackIndex& a, const StackIndex& b) {
       return a._stack_index == b._stack_index;
     }
 
     bool is_invalid() {
-      return _stack_index == _invalid;
-    }
-
-    StackIndex()
-      : _stack_index(_invalid) {
+      return _stack_index == invalid;
     }
   };
 
 private:
-  struct Link : public ArenaObj {
-    Link* next;
+  struct TableEntry;
+  using TableEntryStorage = ArrayWithFreeList<TableEntry, mtNMT>;
+  using TableEntryIndex = typename TableEntryStorage::I;
+
+  TableEntryStorage _entry_storage;
+
+  struct TableEntry {
+    TableEntryIndex next;
     StackIndex stack;
-    Link(Link* next, StackIndex v)
-      : next(next),
-        stack(v) {
-    }
   };
-  StackIndex put(const NativeCallStack& value) {
-    int bucket = value.calculate_hash() % _table_size;
-    Link* link = _table[bucket];
-    while (link != nullptr) {
-      if (value.equals(get(link->stack))) {
-        return link->stack;
-      }
-      link = link->next;
-    }
-    int idx = _stacks.append(value);
-    Link* new_link = new (&_arena) Link(_table[bucket], StackIndex(idx));
-    _table[bucket] = new_link;
-    return new_link->stack;
-  }
 
-  // For storage of the Links
-  Arena _arena;
+  StackIndex put(const NativeCallStack& value);
+
   // Pick a prime number of buckets.
   // 4099 gives a 50% probability of collisions at 76 stacks (as per birthday problem).
   static const constexpr int default_table_size = 4099;
   int _table_size;
```

Comment:
Pre-existing. Both table size and, arguably, the table itself are never changed, since we don't resize. At least size could be const, _table too if you move allocation to init list. (latter is a matter of taste, up to you).